### ISSUE TITLE
Missing CanActivate/Resolve Functionality in Ionic2

### ISSUE THREAD

**Issue by ataraxus** [driftyco/ionic#6148]
#### Short description of the problem:

Need to resolve data before `@Page` gets rendered.

#### What behavior are you expecting?
IMHO: Ionics View callbacks `onPage*` to respect/handle returned promises/observables.

#### Detailed description of the problem:
Hello, 

I'm opening this issue after trying to find a solution/help in the forum, SO and slack. Till now there are  good hints for how to work around this issue, but no definitive solution. 

As for now the decorator `@CanActivate` of Angular2 gets not called within Ionic2. Which opens the question what is the correct way to get data async before a Page gets rendered? This issue was posted [on the forums](https://forum.ionicframework.com/t/what-is-the-equivalent-of-resolve-in-ionic-2/37679/15), but didn't get any "solution".

Apparently Ionic2 [does something](https://github.com/driftyco/ionic-code/blob/c546dcf7dcbf334df349b45db81fbec9718d9fea/2.0.0/js/router.dev.js) with the @CanActivate decorator, but its not documented and i can't figure out what it does exactly.

Nevertheless [this guy](https://webcake.co/page-lifecycle-hooks-in-ionic-2) points out one should use Ionic2 View States instead. His example looks like this:

```
  onPageWillEnter() { 
      return this._service.getComments().then(data => this.comments = data);
  }
```
Which looks like he is expecting Ionic to consider the returned promise, but a [quick glance](https://github.com/driftyco/ionic-code/blob/c546dcf7dcbf334df349b45db81fbec9718d9fea/2.0.0/components/nav/view-controller.js) at Ionics sources reveals (at least I think so) that the returned value is ignored. Hence there is no guarantee that the promise gets resolved before the page gets rendered. During a discussion on #ionic-v2 I created [an example](https://plnkr.co/edit/jF9QAcPwGY97cW9dQ7x0) with onPage* and how it does not perform as needed/expected.

Many people suggested to resolve the data before navigating to the page, which burdens the knowledge which data is needed for the page on the callee. This breaks the nice encapsulation of the Angular2 Component in my opinion.

I'm using 

> Cordova CLI: 6.1.1
> Gulp version:  CLI version 3.9.1
> Gulp local:   Local version 3.9.1
> Ionic Framework Version: 2.0.0-beta.4
> Ionic CLI Version: 2.0.0-beta.24
> Ionic App Lib Version: 2.0.0-beta.14
> OS: Windows 7 SP1
> Node Version: v4.4.2

**Comment by derekdon** [comment]
Any update on this? Not having resolve is the main reason why I haven't migrated to ionic2 yet. I can work around the issue, but would rather wait if it's going to land soon.

**Comment by stefan-1st** [comment]
I would love to see this in the ionic code as well!

**Comment by ataraxus** [comment]
I checked recently with 

```
Ionic Framework Version: 2.0.0-beta.7
Ionic CLI Version: 2.0.0-beta.25
Ionic App Lib Version: 2.0.0-beta.15
OS: Windows 7 SP1
Node Version: v4.4.4
```
no progress so far :(

**Comment by melkir** [comment]
I agree, it would be cool to have this feature.
I'm trying to implement an authenticate system with resticted tab, for the moment I'm listening for onPageWillEnter but it's not really efficient since the page is showed 0,1 sec.

**Comment by Hufschmidt** [comment]
This has been driving me nuts the last few days, so I've gone and implemented this myself.

Since the NavController uses transitions directly instead of the relying on Angular2's (relatively new) beta-router, I simply delayed the transition to wait until both the promise returned by the 'entering-view component' via 'ionCanActivate(enteringView, leavingView)' and the 'leaving-view component' via 'ionCanDeactivate(enteringView: ViewController, leavingView: ViewController)' both have been resolved or cancel the transition if any Promise fails.

See Hufschmidt/ionic@77d8244. Might create a PullRequest...
Currently debating wether to pass the views or the controllers (view.instance) to the ionCan... calls.

**Comment by Hufschmidt** [comment]
I need some feedback from the ionic crew: 
_Are there any (immediate) plans to update/rewrite the current nav-controller?_

I'm asking, because I'd like to improve the current implementation to reject the navigation promise (eg. returned by _nav.setRoot()_) upon rejection of an _ionCan..._ promise. For this I'd like to change the current nav-controller to actually use promise-chaining instead of using the [deferred-promise](https://gist.github.com/flaki/f7c0567b5e104c136dbd) [anti-pattern](https://github.com/petkaantonov/bluebird/wiki/Promise-anti-patterns#the-deferred-anti-pattern), so I don't have to pass _done_ (and _failed_) callbacks around...

**Comment by jgw96** [comment]
@Hufschmidt We are currently doing alot of work around navigation. While it is not a rewrite of NavController, as NavController works perfectly for what it is meant to do, we are adding some very cool functionality that should solve this issue. Thanks!

**Comment by Hufschmidt** [comment]
@jgw96 Thanks for the feedback. With this info I'll  probably not put any more work-hours into improving my current changes to the NavController and rather wait on the awesome things to come, keep up the good work. :smiley:

@ataraxus Yes, those where the workarounds I was talking about, thanks for posting the link though, good idea. :smiley:
Sadly both solutions have their own unique flaws:

**sebaferreras solutiuon:**
While I really like this one, especially for the additional UI feedback thanks to the loading-controller the problems remains that this component is fully loaded and gets all of its (angular and ionic) life-cycle hooks executed regardless of the additional async parameters.
Furthermore, if the async parameters reject, the component would be responsible for restoring the previous navigation/application state. (Which could be as simple as nav.pop() but could also get arbitrarily complex)

**Thomas Withaars solution:**
While this solutions does not suffer from the aboe flaws, now the/each component responsible for navigation has to know about the requirements of every component it can navigate to, circumventing angulars DI for those parameters.

Comparatively I'd say the later  solution (while not the accepted one) is the lesser of both _evils_.

**Comment by akshaycs** [comment]
Is there an update on this now that we have RC0? Thanks!

**Comment by manucorporat** [comment]
RC0 supports ionViewCanLeave and ionViewCanEnter.

**Comment by Luchillo** [comment]
@manucorporat It says so in the changelog, however the [NavController docs](http://ionicframework.com/docs/v2/nightly/api/navigation/NavController/) says nothing about them, how are they used, do they expect true or false? do they expect a promise to resolve or reject?

**Comment by Luchillo** [comment]
I've been playing with it and returning false is the way to stop navigation, however i don't see how to trigger a redirect other than using `setTimeout` to put the redirect last in the js event loop:

```ts
import { Component } from '@angular/core';
import { NavController } from 'ionic-angular';

import { Page2 } from '../page2/page2.ts';

@Component({
  selector: 'page-page1',
  templateUrl: 'page1.html'
})
export class Page1 {
  constructor(public navCtrl: NavController) { }

  ionViewCanEnter() {
    console.log('Can enter.');
    setTimeout(() => {
      this.navCtrl.setRoot(Page2)
    }, 0);
    return false;
  }

}
```

**Comment by Luchillo** [comment]
Ok not so much, that only works at the start, if i just return false in a page that's supposed to be navigated after the root component is been set, it breaks.

**Comment by danishin** [comment]
Is using `ionViewCanEnter` really a valid solution for angular's `resolve`? As far as I can tell, It is equivalent of angular's `CanActivate`, not `resolve`.

**Comment by mohammadshamma** [comment]
This is just awful. I can't conceive how the authentication use case was not fully fleshed out when designing the navigation controller.

**Comment by Luchillo** [comment]
Is still no way to control authentication?

**Comment by ataraxus** [comment]
@Luchillo, @mohammadshamma have a look at the navcontroller `ionViewCanEnter` it does exactly what you are looking for. http://ionicframework.com/docs/v2/nightly/api/navigation/NavController

**Comment by ahoebeke** [comment]
Indeed, thanks @ataraxus for the link. Seems the documentation contains very clear examples and explanations now.
when doing a "push()" on the nav controller, a ".catch()" is needed to intercept the true/false values emitted by the ionViewCanLeave/Enter (use canLeave in same component, or canEnter in the destination page or component).

Haven't had time to actually test this yet, but it's clearly written so hopefully reproducing this will be straightforward and simple. 👍

**Comment by krysalead** [comment]
Hi,

I jump in the discussion as I am also facing the same issue. I am trying to secure pages of my app but if we compare the ionViewCanEnter and canActivate they don't have the same signature which quite annoying if we want to achieve the same goal.

**Comment by mohammadshamma** [comment]
@ataraxus My criticism is for the fact that there is no cross cutting way of doing the check for every page. My understanding is that I might have to define the `ionViewCanEnter` for every page that expects authentication gating.

**Comment by ataraxus** [comment]
@mohammadshamma I think this, as it is always a trade of... either you whitelist or you blacklist. If you dont want to duplicate your auth code, just create a class SecurePage with your ionViewCanEnter and subclass all pages need to be secured. I know its not as common in JS to subclass, I need to get used to it too ;)
